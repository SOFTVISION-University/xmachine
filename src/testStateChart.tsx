import { getShortestPaths } from 'xstate/lib/graph';
import {Segment } from 'xstate/lib/types';

export interface IFixtures {
    [key: string]: any;
}

export const testStateChart = (classInstance: {[key: string]: any}, fixtures: IFixtures = {}, keysToCapture = Object.keys(classInstance)) => {
    const stateMachine = classInstance.getMachine();
    const paths = getShortestPaths(stateMachine)


    Object.keys(paths).forEach(key => {
        const segments = paths[key];
        segments.forEach((segment: Segment) => classInstance.transition(segment.event, fixtures[segment.event as string]))
        const snapshot = keysToCapture.reduce((accumulator, k) => {
             return Object.assign(accumulator, {[k]: classInstance[k]});
        }, {});
        expect(snapshot).toMatchSnapshot(`current state - ${key}`);
    });

} 