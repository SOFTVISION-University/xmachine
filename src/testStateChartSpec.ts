
import { LightMachineStore } from '../examples/mobx/LightMachineStore';
import { testStateChart } from './testStateChart';

describe('State chart', () => {

const lightStore = new LightMachineStore();

    it('should test lightMachine in all possible states', () => {
        testStateChart(lightStore, {}, ['color']);
    });
});