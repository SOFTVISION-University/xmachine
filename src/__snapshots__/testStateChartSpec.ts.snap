// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State chart should test lightMachine in all possible states: current state - "green" 1`] = `
Object {
  "color": "green",
}
`;

exports[`State chart should test lightMachine in all possible states: current state - "red" 1`] = `
Object {
  "color": "green",
}
`;

exports[`State chart should test lightMachine in all possible states: current state - "yellow" 1`] = `
Object {
  "color": "yellow",
}
`;
